### PR TITLE
[gh-pages] End-to-End Visual Control Exercise

### LINKED ISSUE #3217: [gh-pages] End-to-End Visual Control Exercise
[gh-pages] End-to-End Visual Control Exercise

[ Temporary Link ](https://codezerro.github.io/RoboticsAcademy/exercises/AutonomousCars/end_to_end_visual_control/)

### PR DESCRIPTION
**Pull request by codezerro**
[gh-pages] End-to-End Visual Control Exercise

[ Temporary Link ](https://codezerro.github.io/RoboticsAcademy/exercises/AutonomousCars/end_to_end_visual_control/) 

Closes #3217

### CODE REVIEW

File: _pages/exercises/AutonomousCars/end_to_end_visual_control.md [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+standard range, which improves training stability. The <strong>labels</strong> are extracted directly from the CSV files and linked to the appropriate images. To enhance <strong>robustness</strong> and <strong>generalization</strong>, <strong>data augmentation techniques</strong> 
+such as <strong>flipping</strong>, <strong>rotation</strong>, and <strong>brightness adjustment</strong> are applied, 
+creating diverse variations of the training data.
+</p>
+
+#### ii) Dataset Splitting
+
+<p style="text-align:justify">
+The <strong>Simple Circuit dataset</strong> is divided into three sets: <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">training</code> and <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">testing</code>. The training set contains the majority of the data and is used to teach the model to recognize patterns and learn the task. The testing set is the data used during training to fine-tune hyper-parameters, prevent over-fitting, and monitor the model’s performance on unseen data.
```
Comment (dpascualhe) [inline _pages/exercises/AutonomousCars/end_to_end_visual_control.md]: This is a bit misleading. We have two sets (training and testing), not three. Ideally, we would have train, validation and test. In our case, we will test directly against the simulation in RoboticsAcademy, and so the "testing" split in your HuggingFace dataset serves actually as validation. Can you clarify that?

File: _pages/exercises/AutonomousCars/end_to_end_visual_control.md [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+</p>
+
+<p style="text-align:justify;">
+The <strong>Combine Circuit Dataset</strong> consists of <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">training</code> data and <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">adjustment data</code> (used for sharp turns). If needed, these can be combined to create a larger dataset. You can then split the combined dataset into <strong>training</strong> and <strong>testing</strong> sets using an <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">80%-20% ratio</code> with the <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">train_test_split</code> function <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">(from sklearn.model_selection)</code>.
```
Comment (dpascualhe) [inline _pages/exercises/AutonomousCars/end_to_end_visual_control.md]: Same here, it is more correct to say that "You can then split the combined dataset into training and **validation** sets". Again, we don't have a true test set, our test is done directly in simulation.

File: _pages/exercises/AutonomousCars/end_to_end_visual_control.md [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+all_data = train_data + adjustment_data
+
+# Split into 80% train, 20% validation
+train_set, val_set = train_test_split(all_data, test_size=0.2, random_state=42)
+```
+
+#### iii) Model Architecture and Training:
+
+<p style="text-align:justify">
+For this exercise, a deep learning model, such as a Convolutional Neural Network (CNN) like <strong>PilotNet</strong> or <strong>ResNet</strong>, is chosen to perform End-to-End Visual Control. The model consists of multiple convolutional layers that automatically extract meaningful features from raw camera images, followed by fully connected layers that map these features to the predicted outputs, such as linear and angular velocities for the autonomous vehicle.
```
Comment (dpascualhe) [inline _pages/exercises/AutonomousCars/end_to_end_visual_control.md]: I would point here to torchvision pretrained models as a starting point. Also link to PilotNet.

File: _pages/exercises/AutonomousCars/end_to_end_visual_control.md [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+</p>
+<p style="text-align:justify">
+During training, the model learns by comparing its predictions with the correct outputs (ground truth) from the dataset and trying to reduce the difference. This process uses <strong>backpropagation</strong>, which calculates how the model’s internal parameters (weights) should be adjusted to improve accuracy. <strong>Optimization</strong> algorithms like <strong>Adam</strong> are then applied to update these weights step by step. To make the model more reliable and avoid <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">overfitting, techniques such as normalization, dropout, and adjusting</code> the learning rate may also be used.
```
Comment (dpascualhe) [inline _pages/exercises/AutonomousCars/end_to_end_visual_control.md]: "the learning rate may also de **adjusted**"

File: _pages/exercises/AutonomousCars/end_to_end_visual_control.md [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+</p>
+
+#### iv) Validation, Testing, and Evaluation:
+
+<p style="text-align:justify">
+While training, testing dataset is used to monitor the model’s performance on unseen data with the help of the <strong>Loss Validation</strong> graph. This helps ensure that the model is not just memorizing the training data (overfitting). Techniques such as <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">early stopping</code> (ending training when validation performance stops improving) or <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">learning rate</code> scheduling (gradually reducing the learning rate) may be applied to improve results.
+</p>
```
Comment (dpascualhe) [inline _pages/exercises/AutonomousCars/end_to_end_visual_control.md:205]: "While training, **validation** dataset is used to monitor...".

**Review by dpascualhe — CHANGES_REQUESTED** [review]
Remarkable work 😉 The guide is very informative and I think a student could follow up step by step to complete the exercise. I've added some comments, mainly regarding test/validation splits.

**Review by javizqh — DISMISSED** [review]

File: _pages/exercises/AutonomousCars/end_to_end_visual_control.md (hunk 3 above)
Comment (dpascualhe) [reply _pages/exercises/AutonomousCars/end_to_end_visual_control.md]: @codezerro can you address this? it is only change left from my side, otherwise we can merge 😄

Comment (codezerro) [reply _pages/exercises/AutonomousCars/end_to_end_visual_control.md]: @dpascualhe take a look

**Review by dpascualhe — APPROVED** [review]
